src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -618,6 +619,41 @@ public override BoundNode VisitRefTypeOperator(BoundRefTypeOperator node)
             return node.Update(operand, getTypeFromHandle, type);
         }
 
+        private BoundStatement? RewriteFieldOrPropertyInitializer(BoundStatement initializer)
+        {
+            var previousLocals = _additionalLocals;
+            _additionalLocals = ArrayBuilder<LocalSymbol>.GetInstance();
+
+            try
+            {
+                if (initializer.Kind == BoundKind.Block)
+                {
+                    var block = (BoundBlock)initializer;
+
+                    var statement = RewriteExpressionStatement((BoundExpressionStatement)block.Statements.Single(), suppressInstrumentation: true);
+                    Debug.Assert(statement is { });
+                    return block.Update(block.Locals.AddRange(_additionalLocals), block.LocalFunctions, block.HasUnsafeModifier, block.Instrumentation, ImmutableArray.Create(statement));
+                }
+                else
+                {
+                    var statement = RewriteExpressionStatement((BoundExpressionStatement)initializer, suppressInstrumentation: true);
+                    if (statement is null || _additionalLocals.Count == 0)
+                    {
+                        return statement;
+                    }
+                    return new BoundBlock(
```

Comment:
I'm a little worried about the fact that a `Visit` method besides `VisitBlock` has to handle additional locals.
How do we know whether any other `Visit` methods (current or future) also needs to handle that? Is there anything we could do to help remembering to handle?

Reply:
I haven't thought of an easy way to assert that `_additionalLocals` is always at the expected scope. If there are `Visit` methods that represent blocks but are not setting `_additionalLocals`, then there two possibilities. If the block is a top-level block, we should fail the assert in `VisitArrayOrSpanCollectionExpression()` that `_additionalLocals is { }`. Otherwise, `_additionalLocals` will represent the containing block of the actual block. And for the latter case, that should just mean that the inline array temporaries will have longer scope than is strictly necessary.

Reply:
Adding assert to `LocalRewriter.VisitCollectionExpression()` that `_additionalLocals is { }` to increase the chance of catching other cases.